flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/join/DataStreamInnerJoin.scala

```diff
@@ -0,0 +1,287 @@
+
+  override def open(parameters: Configuration): Unit = {
+    LOG.debug(s"Compiling JoinFunction: $genJoinFuncName \n\n " +
+                s"Code:\n$genJoinFuncCode")
+    val clazz = compile(
+      getRuntimeContext.getUserCodeClassLoader,
+      genJoinFuncName,
+      genJoinFuncCode)
+    LOG.debug("Instantiating JoinFunction.")
+    joinFunction = clazz.newInstance()
+
+    // initialize left and right state, the first element of tuple2 indicates how many rows of
+    // this row, while the second element represents the expired time of this row.
+    val tupleTypeInfo = new TupleTypeInfo[JTuple2[Int, Long]](Types.INT, Types.LONG)
+    val leftStateDescriptor = new MapStateDescriptor[Row, JTuple2[Int, Long]](
+      "left", leftType, tupleTypeInfo)
+    val rightStateDescriptor = new MapStateDescriptor[Row, JTuple2[Int, Long]](
+      "right", rightType, tupleTypeInfo)
+    leftState = getRuntimeContext.getMapState(leftStateDescriptor)
+    rightState = getRuntimeContext.getMapState(rightStateDescriptor)
+
+    // initialize timer state
+    val valueStateDescriptor1 = new ValueStateDescriptor[Long]("timervaluestate1", classOf[Long])
+    leftTimer = getRuntimeContext.getState(valueStateDescriptor1)
+    val valueStateDescriptor2 = new ValueStateDescriptor[Long]("timervaluestate2", classOf[Long])
+    rightTimer = getRuntimeContext.getState(valueStateDescriptor2)
+
+    cRowWrapper = new CRowWrappingMultiOuputCollector()
+  }
+
+  /**
+    * Process left stream records
+    *
+    * @param valueC The input value.
+    * @param ctx    The ctx to register timer or get current time
+    * @param out    The collector for returning result values.
+    *
+    */
+  override def processElement1(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    processElement(valueC, ctx, out, leftTimer, leftState, rightState, true)
+  }
+
+  /**
+    * Process right stream records
+    *
+    * @param valueC The input value.
+    * @param ctx    The ctx to register timer or get current time
+    * @param out    The collector for returning result values.
+    *
+    */
+  override def processElement2(
+      valueC: CRow,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#Context,
+      out: Collector[CRow]): Unit = {
+
+    processElement(valueC, ctx, out, rightTimer, rightState, leftState, false)
+  }
+
+
+  /**
+    * Called when a processing timer trigger.
+    * Expire left/right records which are expired in left and right state.
+    *
+    * @param timestamp The timestamp of the firing timer.
+    * @param ctx       The ctx to register timer or get current time
+    * @param out       The collector for returning result values.
+    */
+  override def onTimer(
+      timestamp: Long,
+      ctx: CoProcessFunction[CRow, CRow, CRow]#OnTimerContext,
+      out: Collector[CRow]): Unit = {
+
+    if (stateCleaningEnabled && leftTimer.value == timestamp) {
```

Comment:
It seems if `stateCleaningEnabled = false`, there will be no timer registered, right?

Reply:
yes~

Reply:
Then the `onTimer()` method will be invoked only if `stateCleaningEnabled = true`. The condition can be removed here.